pkg/operator/controller/extension/virtualcluster/reconciler.go

```diff
@@ -0,0 +1,316 @@
+	conditions := NewVirtualClusterConditions(r.Clock, extension.Status)
+	if !controllerutil.ContainsFinalizer(extension, operatorv1alpha1.FinalizerName) {
+		log.Info("Adding finalizer")
+		if err := controllerutils.AddFinalizers(ctx, r.RuntimeClient, extension, operatorv1alpha1.FinalizerName); err != nil {
+			return fmt.Errorf("failed to add finalizer: %w", err)
+		}
+	}
+
+	log.Info("Reconciling extension virtual resources", "name", extension.Name)
+	err := r.reconcileVirtualClusterResources(ctx, log, gardenClient, extension)
+	if err != nil {
+		conditions.virtualClusterReconciled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.virtualClusterReconciled, gardencorev1beta1.ConditionFalse, ConditionReconcileFailed, err.Error())
+	} else {
+		conditions.virtualClusterReconciled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.virtualClusterReconciled, gardencorev1beta1.ConditionTrue, ConditionReconcileSuccess, fmt.Sprintf("Extension %q has been reconciled successfully", extension.Name))
+	}
+
+	return errors.Join(err, r.updateExtensionStatus(ctx, log, extension, conditions))
+}
+
+func (r *Reconciler) reconcileVirtualClusterResources(ctx context.Context, log logr.Logger, gardenClient client.Client, extension *operatorv1alpha1.Extension) error {
+	// return early if we do not have to make a deployment
+	if extension.Spec.Deployment == nil ||
+		extension.Spec.Deployment.ExtensionDeployment == nil ||
+		extension.Spec.Deployment.ExtensionDeployment.Helm == nil {
+		return r.deleteVirtualClusterResources(ctx, log, gardenClient, extension)
+	}
+
+	if err := r.reconcileControllerDeployment(ctx, gardenClient, extension); err != nil {
+		err := fmt.Errorf("failed to reconciler controller deployment: %w", err)
+		return err
+	}
+	r.Recorder.Event(extension, corev1.EventTypeNormal, "Reconciliation", "ControllerDeployment created successfully")
+
+	if err := r.reconcileControllerRegistration(ctx, gardenClient, extension); err != nil {
+		err := fmt.Errorf("failed to reconciler controller registration: %w", err)
+		return err
+	}
+	r.Recorder.Event(extension, corev1.EventTypeNormal, "Reconciliation", "ControllerRegistration created successfully")
+	return nil
+}
+
+func (r *Reconciler) reconcileControllerDeployment(ctx context.Context, gardenClient client.Client, extension *operatorv1alpha1.Extension) error {
+	ctrlDeploy := &gardencorev1.ControllerDeployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: extension.Name,
+		},
+	}
+
+	deployMutateFn := func() error {
+		ctrlDeploy.Helm = &gardencorev1.HelmControllerDeployment{
+			Values:        extension.Spec.Deployment.ExtensionDeployment.Values,
+			OCIRepository: extension.Spec.Deployment.ExtensionDeployment.Helm.OCIRepository,
+		}
+		return nil
+	}
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, gardenClient, ctrlDeploy, deployMutateFn); err != nil {
+		return fmt.Errorf("failed to create or update ControllerInstallation: %w", err)
```

Comment:
A `ControllerDeployment` is created here.
```suggestion
		return fmt.Errorf("failed to create or update ControllerDeployment: %w", err)
```